apps/hannk/interpreter/interpreter.cpp

```diff
@@ -17,19 +23,127 @@ Interpreter::~Interpreter() {
 
 namespace {
 
-class AllocateAll : public OpVisitor {
-    void visit(OpGroup *g) {
+// TODO: maybe move this to a separate file? Not sure if it's complex enough to be worthy or not.
+class TensorVisitor : public OpVisitor {
+    virtual void visit_tensor(const TensorPtr &t) = 0;
+
+    void visit(OpGroup *g) override {
         for (int i = 0; i < g->op_count(); i++) {
+            op_index_++;
```

Comment:
> Depending on this index for lifetime checking seems sketchy to me. It would be easy for this to be invalidated...

Can you expand on this comment some? What are some ways it could be invalidated? What better approaches would catch those invalidations? (perhaps examination of producer/consumer?) This approach is simpleminded but seemed adequate.
 
> Also, I'm not sure breaking this into a base and derived visitor is worth it, given how tightly coupled this is and there is only one derived user currently.

There were several when I was working on this, and so this is the vestigial organ left behind. I'll collapse then together.

Reply:
whoops -- actually, there are two derived classes (FindAllocatableTensors and VerifyAllAllocated, though the latter is debug-only), so I'm going to leave it for now.